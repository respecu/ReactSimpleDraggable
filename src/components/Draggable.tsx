import React, { Component, createRef } from "react";
import './styles.css';

interface DraggableState {
    top: number,
    left: number,
    dragging: boolean
}

export default class Draggable extends Component<any, DraggableState> {
    private _draggable: any = createRef();
    
    constructor(props: any) {
        super(props);

        this.state = {
            top: 100,
            left: 100,
            dragging: false,
        }
    }

    componentDidUpdate(props: any, state: any) {
        if (this.state.dragging && !state.dragging) {
          document.addEventListener('mousemove', this.onMouseMove)
          document.addEventListener('mouseup', this.onMouseUp)
        } else if (!this.state.dragging && state.dragging) {
          document.removeEventListener('mousemove', this.onMouseMove)
          document.removeEventListener('mouseup', this.onMouseUp)
        }
    }

    onMouseMove = (e: any) => {
        if (!this.state.dragging) {
            return;
        }

        let left = e.x - this._draggable.offsetWidth / 2;
        let top = e.y - this._draggable.offsetHeight / 2;

        if(left > window.innerWidth) {
            left = window.innerWidth - this._draggable.clientWidth;
        }
        else if(left < 0) {
            left = 0;
        }

        if(top > window.innerHeight) {
            top = window.innerHeight - this._draggable.clientHeight;
        }
        else if(top < 0) {
            top = 0;
        }

        // console.log(left, top, window.innerWidth, window.innerHeight);
        // console.log(this._draggable);

        this.setState({
            left,
            top
        })

        e.stopPropagation();
        e.preventDefault();
      }
    
      onMouseUp = (e: any) => {
        this.setState({dragging: false});
        e.stopPropagation();
        e.preventDefault();
      }
    
      onMouseDown = (e: any) => {
        if (e.button !== 0) {
            return;
        }

        const left = e.clientX - this._draggable.offsetWidth / 2;
        const top = e.clientY - this._draggable.offsetHeight / 2;

        // console.log(left, top);

        this.setState({
            dragging: true,
            left,
            top,
        });
    
        e.stopPropagation();
        e.preventDefault();
      }

    render() {
        const { top, left } = this.state;

        return (
            <div className='draggable' style={{top: top, left: left}}
            ref={c => this._draggable = c}
            onMouseDown={this.onMouseDown}
            >
                {this.props.children}
            </div>
        )
    }
}