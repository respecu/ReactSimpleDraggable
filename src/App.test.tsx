import { render } from '@testing-library/react';
import Enzyme, { shallow } from 'enzyme';
import ReactSixteenAdapter from 'enzyme-adapter-react-16';
import React from 'react';
import App from './App';
import Box from './components/Box';
import Draggable from './components/Draggable';

Enzyme.configure({ adapter: new ReactSixteenAdapter() });

function mouseMove(x: number, y: number) {
  const evt = document.createEvent('MouseEvents');
  evt.initMouseEvent("mousemove", true, true, window, 0, 0, 0, x, y, false, false, false, false, 0, null);
  document.dispatchEvent(evt);
  return evt;
}

test('renders learn react link', () => {
  render(<App />);
  // const linkElement = screen.getByText(/learn react/i);
  // expect(linkElement).toBeInTheDocument();
});

test('drag test', () => {
  const draggable = shallow(
    <Draggable>
      <Box />
    </Draggable>
  );

  draggable.simulate("onMouseDown");

  mouseMove(115, 319);
  mouseMove(214, 325);
  mouseMove(225, 356);
  mouseMove(255, 375);
  mouseMove(999999, 999999);

  draggable.simulate("onMouseUp");
})