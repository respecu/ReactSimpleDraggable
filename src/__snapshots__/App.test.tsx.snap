// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drag test 1`] = `
<div
  className="draggable"
  onMouseDown={[Function]}
  style={
    Object {
      "left": 100,
      "top": 100,
    }
  }
>
  <div
    className="box"
  />
</div>
`;
